### ISSUE TITLE
Export to NCNN format no longer works with Ultralytics 8.3.71 and torch 2.6.0+cpu on Raspberry Pi 4

### ISSUE THREAD

**Issue by EdjeElectronics** [ultralytics/ultralytics#19091]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

The latest version of `ultralytics 8.3.71` in conjunction with `torch-2.6.0+cpu` causes the `yolo export model=yolo11n.pt format=ncnn` to longer work. It results in the following output, and the model does not get converted to NCNN format. If you downgrade the libraries to `ultralytics 8.3.70` and `torch-2.5.1+cpu`, it does work.

Can you please resolve this issue? Is there any way I can help?

```
yolo export model=my_model.pt format=ncnn
Ultralytics 8.3.71     Python-3.11.2 torch-2.6.0+cpu CPU (Cortex-A72)
YOLO11n summary (fused): 238 layers, 2,583,127 parameters, 0 gradients, 6.3 GFLOPs

PyTorch: starting from 'my_model.pt' with input shape (1, 3, 480, 480) BCHW and output shape(s) (1, 9, 4725) (5.2 MB)

TorchScript: starting export with torch 2.6.0+cpu...
Illegal instruction
```

![Image](https://github.com/user-attachments/assets/9794a459-f174-4f16-8fa8-662c3dc61909)

Thanks to the Ultralytics team for all your great work on this library!

### Environment

(venv) evan@raspberrypi:~/yolo/yolo3 $ yolo checks
Ultralytics 8.3.71 🚀 Python-3.11.2 torch-2.6.0+cpu CPU (Cortex-A72)
Setup complete ✅ (4 CPUs, 7.6 GB RAM, 12.0/27.8 GB disk)

OS                  Linux-6.6.62+rpt-rpi-v8-aarch64-with-glibc2.36
Environment         Linux
Python              3.11.2
Install             pip
RAM                 7.63 GB
Disk                12.0/27.8 GB
CPU                 Cortex-A72
CPU count           4
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 2.1.1<=2.1.1,>=1.23.0
matplotlib          ✅ 3.10.0>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.1>=1.4.1
torch               ✅ 2.6.0>=1.8.0
torch               ✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 6.1.1
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0

### Minimal Reproducible Example

**Steps to reproduce**

This is all run on a Raspberry Pi 4b 8GB using 64-bit Raspberry Pi OS version 12 (Bookworm). The OS is freshly installed on an SD cardusing Raspberry Pi Imager.

Open a terminal and run the following commands:

1. `sudo apt update && sudo apt upgrade -y`
2. `mkdir yolo && cd yolo`
3. `python3 -m venv venv`
4. `source venv/bin/activate`
5. `pip install ultralytics ncnn`
6. `yolo export model=yolo11n.pt format=ncnn`

The error will occur upon running that last command. The error no longer occurs if you downgrade using `pip install ultralytics==8.3.70 torch==2.5.1`.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @EdjeElectronics, thank you for bringing this to our attention and for your detailed report 🚀! We appreciate your support and are here to assist.

If this is indeed a 🐛 Bug Report, could you please ensure you have provided a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate this issue further? From your concise steps to reproduce the error, it seems clear, but any additional details on your setup, exact inputs, or logs (if available) would be great.

In the meantime, we recommend trying the following:
1. **Upgrade:** Ensure that you are running the latest version of the `ultralytics` package. You can upgrade using:

```bash
pip install -U ultralytics
```

2. **Environment Validation:** YOLO works in various verified environments. If possible, test your export command in one of the environments below to help isolate the issue:
   - **Notebooks** with free GPU: 
     <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> 
     <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> 
     <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
   - **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
   - **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
   - **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

3. **Suggestions:** Since the issue involves a specific combination of `torch-2.6.0+cpu` and `ultralytics 8.3.71`, you might temporarily revert to the working versions (`ultralytics 8.3.70` and `torch-2.5.1+cpu`) for tasks requiring NCNN exports until the issue is resolved.

Additionally, feel free to connect with our vibrant Ultralytics community for more insights and real-time discussions:
- Join us on [Discord](https://discord.com/invite/ultralytics) 🎧
- Participate in threads on our [Subreddit](https://reddit.com/r/Ultralytics)
- Engage in discussions on [Discourse](https://community.ultralytics.com/)

Lastly, our CI badges below reflect the current operational status of our tests:
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

An Ultralytics engineer will take a closer look at your issue soon. Thank you for your patience and for contributing to the community 😊!

**Comment by Y-T-G** [comment]
It works for me. Seems like an issue with Raspberry Pi's   PyTorch.

Does prediction work with torch 2.6?

You should open an issue on PyTorch repo.

**Comment by younesslanda** [comment]
I am facing kind ot the same issue, using yolo12n with ncnn on a raspberry pi 5.

```python
from ultralytics import YOLO

# Load a COCO-pretrained YOLO12n model
model = YOLO("yolo12n.pt")

model.export(format="ncnn")

ncnn_model = YOLO("yolo12n_ncnn_model")

results = ncnn_model("https://ultralytics.com/images/bus.jpg")
```

The code above takes unusual long time to execute: 900 ms for one frame and it outputs `no detection`


Plus, if I execute this part `results = ncnn_model("https://ultralytics.com/images/bus.jpg")` for a couple of times (in a jupyter notebook cell), it crashes with `Segmentation fault`

**Comment by glenn-jocher** [comment]
@younesslanda thanks for sharing this issue. YOLO12 models are not currently recommended due to known instability and performance concerns, especially on CPU devices like Raspberry Pi. We suggest using YOLO11n instead, which is optimized for edge deployment and shows significantly better speed/accuracy tradeoffs. 

For NCNN deployment, please follow the [YOLO11 documentation](https://docs.ultralytics.com/models/yolo11/) and ensure you're using the latest Ultralytics pip package. If issues persist after switching to YOLO11, we recommend creating a minimal reproducible example as described in our [bug report guidelines](https://github.com/ultralytics/ultralytics/issues/new/choose).

**Comment by Karthik43-sys** [comment]
We are using this yolo11n.pt file but we cant convert this pytorch to ncnn please help me with the correct versions and commands

**Comment by Karthik43-sys** [comment]
We are using raspberry pi model 4B

**Comment by Karthik43-sys** [comment]
The error message is illegal instruction

**Comment by Y-T-G** [comment]
You should downgrade PyTorch. It's a PyTorch bug.

**Comment by Karthik43-sys** [comment]
pip install ultralytics==8.3.70 torch==2.5.0 torchvision==0.20", we give this command but the error still exist "illegal instruction"

**Comment by Y-T-G** [comment]
Try 2.3.1

**Comment by mmartignago** [comment]
Same issue with ONNX
![Image](https://github.com/user-attachments/assets/76ee189e-209f-4ac9-bc71-deb6e781ab87)

**Comment by glenn-jocher** [comment]
Thank you for reporting this issue. The "Illegal instruction" error appears related to PyTorch compatibility on ARM architectures like Raspberry Pi. We recommend: 

1. Confirming your PyTorch installation matches Raspberry Pi's ARMv8 architecture requirements
2. Creating a minimal reproducible example per our [bug report guidelines](https://docs.ultralytics.com/help/minimum-reproducible-example/) 
3. Opening a GitHub issue with your full environment details from `yolo checks`

This will help us investigate further. For immediate needs, you might consider using our recommended [YOLO11 edge-optimized models](https://docs.ultralytics.com/models/yolo11/#edge-devices).

**Comment by JugPanda** [comment]
@Karthik43-sys
> pip install ultralytics==8.3.70 torch==2.5.0 torchvision==0.20", we give this command but the error still exist "illegal instruction"


@glenn-jocher The command **below** worked for me, which allowed me to convert a yolo11n model to ncnn format: 

> **pip install ultralytics==8.3.70 torch==2.5.0 torchvision==0.20.0 --no-cache-dir**

I found this command in the comment section of @EdjeElectronics's YouTube video: [https://www.youtube.com/watch?v=z70ZrSZNi-8](url)

![Image](https://github.com/user-attachments/assets/2ec85b9c-bf4d-46f4-891a-a513084d271e)

**Comment by mmartignago** [comment]
Yes thank you that worked

**Comment by glenn-jocher** [comment]
Glad to hear it worked! The `--no-cache-dir` flag can help avoid installation conflicts on ARM devices like Raspberry Pi. This appears to be a PyTorch compatibility issue with ARM architectures that we're monitoring. For others facing similar issues, our [Raspberry Pi deployment guide](https://docs.ultralytics.com/guides/raspberry-pi/) provides additional optimization tips.

**Comment by bryanmmm** [comment]
Hello there! I have tried to follow these recommendations for the error, but it seems that  there is an ERROR: No matching distribution found for torch==2.5.0 or any version other than 2.6.0 and up. Has anyone run into this as well? If so any way around it?

**Comment by onuralpszr** [comment]
> Hello there! I have tried to follow these recommendations for the error, but it seems that there is an ERROR: No matching distribution found for torch==2.5.0 or any version other than 2.6.0 and up. Has anyone run into this as well? If so any way around it?

Hello, Can you try with latest package version.

**Comment by bryanmmm** [comment]
> > Hello there! I have tried to follow these recommendations for the error, but it seems that there is an ERROR: No matching distribution found for torch==2.5.0 or any version other than 2.6.0 and up. Has anyone run into this as well? If so any way around it?
> 
> Hello, Can you try with latest package version.

I was able to figure the sweet spot, torch==2.7.0 torchvision==0.21 ultralytics=8.3.115

**Comment by onuralpszr** [comment]
> > > Hello there! I have tried to follow these recommendations for the error, but it seems that there is an ERROR: No matching distribution found for torch==2.5.0 or any version other than 2.6.0 and up. Has anyone run into this as well? If so any way around it?
> > 
> > 
> > Hello, Can you try with latest package version.
> 
> I was able to figure the sweet spot, torch==2.7.0 torchvision==0.21 ultralytics=8.3.115


I was able to export by using latest one as well. You shouldn't need to look for specific version of packages.  

```bash
pip list | grep -E "torch|ncnn|pnnx|ultralytics"

ncnn                         1.0.20260114
pnnx                         20260112
torch                         2.9.1
torchvision               0.24.1
ultralytics                   8.4.14    
ultralytics-thop          2.0.18
```

Export Report

```bash
❯ yolo export model=yolo26n.pt format=ncnn
Ultralytics 8.4.14 🚀 Python-3.13.12 torch-2.9.1 CPU
WARNING ⚠️ NCNN export does not support end2end models, disabling end2end branch.
YOLO26n summary (fused): 146 layers, 2,562,496 parameters, 0 gradients, 5.4 GFLOPs

PyTorch: starting from 'yolo26n.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (5.3 MB)

NCNN: starting export with NCNN 1.0.20260114 and PNNX 20260112...
pnnxparam = yolo26n_ncnn_model/model.pnnx.param
pnnxbin = yolo26n_ncnn_model/model.pnnx.bin
pnnxpy = yolo26n_ncnn_model/model_pnnx.py
pnnxonnx = yolo26n_ncnn_model/model.pnnx.onnx
ncnnparam = yolo26n_ncnn_model/model.ncnn.param
ncnnbin = yolo26n_ncnn_model/model.ncnn.bin
ncnnpy = yolo26n_ncnn_model/model_ncnn.py
fp16 = 0
optlevel = 2
device = cpu
inputshape = [1,3,640,640]f32
inputshape2 =
customop =
moduleop =
get inputshape from traced inputs
inputshape = [1,3,640,640]f32
############# pass_level0
inline module = torch.nn.modules.linear.Identity
inline module = ultralytics.nn.modules.block.Attention
inline module = ultralytics.nn.modules.block.Bottleneck
inline module = ultralytics.nn.modules.block.C2PSA
inline module = ultralytics.nn.modules.block.C3k
inline module = ultralytics.nn.modules.block.C3k2
inline module = ultralytics.nn.modules.block.PSABlock
inline module = ultralytics.nn.modules.block.SPPF
inline module = ultralytics.nn.modules.conv.Concat
inline module = ultralytics.nn.modules.conv.Conv
inline module = ultralytics.nn.modules.conv.DWConv
inline module = ultralytics.nn.modules.head.Detect
inline module = torch.nn.modules.linear.Identity
inline module = ultralytics.nn.modules.block.Attention
inline module = ultralytics.nn.modules.block.Bottleneck
inline module = ultralytics.nn.modules.block.C2PSA
inline module = ultralytics.nn.modules.block.C3k
inline module = ultralytics.nn.modules.block.C3k2
inline module = ultralytics.nn.modules.block.PSABlock
inline module = ultralytics.nn.modules.block.SPPF
inline module = ultralytics.nn.modules.conv.Concat
inline module = ultralytics.nn.modules.conv.Conv
inline module = ultralytics.nn.modules.conv.DWConv
inline module = ultralytics.nn.modules.head.Detect

----------------

############# pass_level1
############# pass_level2
############# pass_level3
############# pass_level4
############# pass_level5
############# pass_ncnn
NCNN: export success ✅ 5.9s, saved as 'yolo26n_ncnn_model' (9.4 MB)

Export complete (6.0s)
Results saved to /Users/onuralp/Github/ultralytics-develop
Predict:         yolo predict task=detect model=yolo26n_ncnn_model imgsz=640
Validate:        yolo val task=detect model=yolo26n_ncnn_model imgsz=640 data=/home/lq/codes/ultralytics/ultralytics/cfg/datasets/coco.yaml
Visualize:       https://netron.app
💡 Learn more at https://docs.ultralytics.com/modes/export

```

Prediction 

```bash
❯ yolo predict task=detect model=yolo26n_ncnn_model imgsz=640
Ultralytics 8.4.14 🚀 Python-3.13.12 torch-2.9.1 CPU (Apple M2 Pro)
Loading yolo26n_ncnn_model for NCNN inference...

image 1/2 /ultralytics/ultralytics/assets/bus.jpg: 640x640 4 persons, 1 bus, 24.3ms
image 2/2 /ultralytics/ultralytics/assets/zidane.jpg: 640x640 2 persons, 1 tie, 18.3ms
Speed: 3.0ms preprocess, 21.3ms inference, 1.7ms postprocess per image at shape (1, 3, 640, 640)
```

**Comment by kimjihobf-dotcom** [comment]
> > > Hello there! I have tried to follow these recommendations for the error, but it seems that there is an ERROR: No matching distribution found for torch==2.5.0 or any version other than 2.6.0 and up. Has anyone run into this as well? If so any way around it?
> > 
> > 
> > Hello, Can you try with latest package version.
> 
> I was able to figure the sweet spot, torch==2.7.0 torchvision==0.21 ultralytics=8.3.115

This doesn’t work for me, it says that torch and torchvision versions clash/are not compatible. Is there any other versions that work that aren’t 2.5.1? Please help me

**Comment by glenn-jocher** [comment]
@kimjihobf-dotcom that usually means the pinned `torch`/`torchvision` combo does not have matching wheels for your exact Python/Raspberry Pi setup, so please don’t mix arbitrary version pins; since this is a closed stale issue, please open a new issue with your full text traceback and `yolo checks` output, and first retry from a clean venv using the [Raspberry Pi NCNN workflow](https://docs.ultralytics.com/guides/raspberry-pi/) with:

```bash
python3 -m venv venv
source venv/bin/activate
pip install -U pip
pip install -U ultralytics[export]
yolo export model=yolo26n.pt format=ncnn
```

If it still fails, also include the output of `python -c "from ultralytics.utils.checks import collect_system_info; collect_system_info()"` from that same environment so we can check for package/path mismatches.
